File: PhotographerApp.Infrastructure/Services/Business/ContactService.cs
using Microsoft.EntityFrameworkCore;
using PhotographerApp.Core.DTOs;
using PhotographerApp.Core.Entities;
using PhotographerApp.Core.Interfaces;
using PhotographerApp.Infrastructure.Data;

namespace PhotographerApp.Infrastructure.Services.Business;

public class ContactService : IContactService
{
    private readonly ApplicationDbContext _context;

    public ContactService(ApplicationDbContext context)
    {
        _context = context;
    }

    public async Task SubmitAsync(SubmitContactMessageRequest request, CancellationToken cancellationToken = default)
    {
        _context.ContactMessages.Add(new ContactMessage
        {
            Name = request.Name,
            Email = request.Email,
            Subject = request.Subject,
            Message = request.Message,
        });

        await _context.SaveChangesAsync(cancellationToken);
    }

    public async Task<PaginatedResponse<ContactMessageDto>> GetMessagesAsync(int pageNumber = 1, int pageSize = 20, CancellationToken cancellationToken = default)
    {
        var query = _context.ContactMessages.OrderByDescending(m => m.CreatedAt);

        var totalCount = await query.CountAsync(cancellationToken);

        var messages = await query
            .Skip((pageNumber - 1) * pageSize)
            .Take(pageSize)
            .Select(m => new ContactMessageDto
            {
                Id = m.Id,
                Name = m.Name,
                Email = m.Email,
                Phone = m.Phone,
                Subject = m.Subject,
                Message = m.Message,
                EventType = m.EventType,
                EventDate = m.EventDate,
                IsRead = m.IsRead,
                CreatedAt = m.CreatedAt,
            })
            .ToListAsync(cancellationToken);

        return new PaginatedResponse<ContactMessageDto>
        {
            Items = messages,
            PageNumber = pageNumber,
            PageSize = pageSize,
            TotalCount = totalCount,
            TotalPages = (int)Math.Ceiling(totalCount / (double)pageSize),
        };
    }

    public async Task<bool> SetReadAsync(int id, bool isRead, CancellationToken cancellationToken = default)
    {
        var message = await _context.ContactMessages.FindAsync(new object[] { id }, cancellationToken);
        if (message == null) return false;

        message.IsRead = isRead;
        await _context.SaveChangesAsync(cancellationToken);
        return true;
    }

    public async Task<bool> DeleteAsync(int id, CancellationToken cancellationToken = default)
    {
        var message = await _context.ContactMessages.FindAsync(new object[] { id }, cancellationToken);
        if (message == null) return false;

        message.IsDeleted = true;
        await _context.SaveChangesAsync(cancellationToken);
        return true;
    }
}

File: src/contact/contact.service.ts
import { Injectable } from '@nestjs/common';
import { PrismaService } from '../prisma/prisma.service.js';
import { PaginatedResponse } from '../common/dto/api-response.dto.js';
import { SubmitContactMessageRequestDto } from './dto/submit-contact-message-request.dto.js';
import { ContactMessageDto } from './dto/contact-message-response.dto.js';

@Injectable()
export class ContactService {
  constructor(private readonly prisma: PrismaService) {}

  async submit(request: SubmitContactMessageRequestDto): Promise<void> {
    await this.prisma.contactMessage.create({
      data: {
        name: request.name,
        email: request.email,
        subject: request.subject,
        message: request.message,
        isRead: false,
        isDeleted: false,
        createdAt: new Date(),
        updatedAt: new Date(),
      },
    });
  }

  async getMessages(pageNumber = 1, pageSize = 20): Promise<PaginatedResponse<ContactMessageDto>> {
    const where = { isDeleted: false };
    const [totalCount, messages] = await Promise.all([
      this.prisma.contactMessage.count({ where }),
      this.prisma.contactMessage.findMany({
        where,
        orderBy: { createdAt: 'desc' },
        skip: (pageNumber - 1) * pageSize,
        take: pageSize,
      }),
    ]);

    return new PaginatedResponse({
      items: messages.map((m) => ({
        id: m.id,
        name: m.name,
        email: m.email,
        phone: m.phone,
        subject: m.subject,
        message: m.message,
        eventType: m.eventType,
        eventDate: m.eventDate,
        isRead: m.isRead,
        createdAt: m.createdAt,
      })),
      pageNumber,
      pageSize,
      totalCount,
    });
  }

  async setRead(id: number, isRead: boolean): Promise<boolean> {
    const message = await this.prisma.contactMessage.findFirst({ where: { id, isDeleted: false } });
    if (!message) return false;

    await this.prisma.contactMessage.update({ where: { id }, data: { isRead, updatedAt: new Date() } });
    return true;
  }

  async delete(id: number): Promise<boolean> {
    const message = await this.prisma.contactMessage.findFirst({ where: { id, isDeleted: false } });
    if (!message) return false;

    await this.prisma.contactMessage.update({ where: { id }, data: { isDeleted: true, updatedAt: new Date() } });
    return true;
  }
}
